### PR TITLE
If the requested slot is empty, get the block header for the latest block.

### LINKED ISSUE #3071: API /eth/v1/beacon/headers/[finalized|head] return 404
`/eth/v1/beacon/headers/{block_id}` returns 404 for `head` and `finalized` rather than the expected headers.

For the second and third queries below I expect to see block headers, as per the interface documentation, rather than the 404 message:

```
ben@metalalbert:~/teku$ curl localhost:5051/eth/v1/beacon/headers/genesis
{"data":{"root":"0xebe49843e96e538592c9e1976e155e9057c251ed9312eba517afce3ca1a0660c","canonical":true,"header":{"message":{"slot":"0","proposer_index":"0","parent_root":"0x0000000000000000000000000000000000000000000000000000000000000000","state_root":"0xaf6aafa94dcc22a5cf0253c4f5dd886397900034bf7c149c54635b537f34f64e","body_root":"0xebe49843e96e538592c9e1976e155e9057c251ed9312eba517afce3ca1a0660c"},"signature":"0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"}}}

ben@metalalbert:~/teku$ curl localhost:5051/eth/v1/beacon/headers/head
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>404 - Page not found</title>
</head>
<body>
<p>The page you have requested doesn't seem to be one we were expecting.</p>
<P>In better news, you can go and check <a href="/swagger-ui">swagger</a> to hopefully find what you need.</P>
</body>
</html>

ben@metalalbert:~/teku$ curl localhost:5051/eth/v1/beacon/headers/finalized
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>404 - Page not found</title>
</head>
<body>
<p>The page you have requested doesn't seem to be one we were expecting.</p>
<P>In better news, you can go and check <a href="/swagger-ui">swagger</a> to hopefully find what you need.</P>
</body>
</html>

ben@metalalbert:~/teku$ curl localhost:5051/eth/v1/beacon/headers/foobarbaz
{"status":400,"message":"Invalid block: foobarbaz"}
```

It looks like #3011 implemented this, but something is broken.

Version: `teku/v0.12.12-dev-5cefb9a2/linux-x86_64/-privatebuild-openjdk64bitservervm-java-13`

### PR DESCRIPTION
**Pull request by rolfyone**
- Previously an empty head slot would result in a 404 (not found) error.

 - It was possible for the slot of the finalized epoch to be empty, which would result in 404.

fixes #3071

Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

**Comment by ajsutton** [comment]
I suspect if a specific slot is requested we should ensure we only return blocks from that slot. Just  `head` and `finalized` would return the block before.  We've had bug reports about the older APIs unexpectedly returning blocks from earlier slots before.

**Comment by cemozerr** [comment]
The https://github.com/ConsenSys/teku/pull/3054 PR already fixes this. Can we approve and merge that instead?

To be fair, if we're only going to return **blocks in effect** for `head` and `finalized`, and **blocks at slot** for all other slot specific requests, that PR is going to require a major re-write.

That is due to the fact that regardless of getting a `head` or let's say `64` (slot) as the parameter, currently, we convert it to the corresponding slot and get the block in effect for that slot. Both for `/headers` and `/blocks` endpoints.

File: data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java [hunk 1]
```diff
@@ -196,24 +199,14 @@ public GetForkResponse getForkInfo() {
...
-        .getBlockByBlockRoot(blockRoot)
-        .thenApply(maybeBlock -> maybeBlock.map(block -> new BlockHeader(block, false)));
+    return defaultBlockSelectorFactory
+        .defaultBlockSelector(slotParameter)
+        .getBlock()
+        .thenApply(
+            blockList ->
+                blockList.isEmpty()
+                    ? Optional.empty()
+                    : Optional.of(new BlockHeader(blockList.get(0), true)));
```
Comment (ajsutton) [inline data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java]: We could have a default method `getSingleBlock` in the `BlockSelector` interface that does this conversion from list to `Optional`.  Could just do a `blockList.stream().findFirst()` so simplify it.

File: networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/core/Eth2OutgoingRequestHandlerTest.java [hunk 2]
```diff
@@ -280,7 +282,7 @@ public void shouldCompleteExceptionallyWhenClosedWithTruncatedMessage() {
 
     asyncRequestRunner.executeQueuedActions();
 
-    complete();
+    assertThrows(IllegalReferenceCountException.class, this::complete);
```
Comment (ajsutton) [inline networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/core/Eth2OutgoingRequestHandlerTest.java]: We shouldn't be expecting this and it probably shouldn't be in this change.  Need to work out why it's being released twice and fix that in a different PR. :)

File: data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java (hunk 1 above)
Comment (ajsutton) [reply data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java]: Also I'd have expected the handler to convert the String slotParameter to a BlockSelector and then pass the BlockSelector into this method.  That way the handler is responsible for interpreting the request and the provider has the business logic.  With it this way the provider is interpreting the request and doing the business logic.

File: networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/core/Eth2OutgoingRequestHandlerTest.java (hunk 2 above)
Comment (rolfyone) [reply networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/core/Eth2OutgoingRequestHandlerTest.java]: alright but it blocks every build i do, so basically i cant get a green build on my machine before i push

Comment (ajsutton) [reply networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/rpc/core/Eth2OutgoingRequestHandlerTest.java]: I don't think this is the right fix either - complete exceptionally is different to throws exception.  I wouldn't have expected to get this exception from a call to `complete`.

**Review by ajsutton — APPROVED** [review]
Agreed to go with the current approach of passing the string down to the provider.  LGTM.

**Pull request merged**